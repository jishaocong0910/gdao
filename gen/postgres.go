package gen

import (
	"database/sql"
	"fmt"
	"strings"

	"github.com/jishaocong0910/gdao"
	_ "github.com/lib/pq"
)

type postgresGenerator struct {
	c        Config
	db       *sql.DB
	schema   string
	database string
}

func (g postgresGenerator) getTableInfo(table string) (bool, []*field, string) {
	var (
		exists       bool
		fields       []*field
		tableComment string
	)

	rows := mustReturn(g.db.Query("SELECT i.column_name,i.udt_name,p.attndims,p.description FROM information_schema.columns i JOIN (SELECT n.nspname,c.relname,a.attname,a.attndims,d.description FROM pg_namespace n JOIN pg_class c ON n.oid = c.relnamespace JOIN pg_attribute a ON a.attrelid = c.oid LEFT JOIN pg_description d ON d.objoid = c.oid AND d.objsubid = a.attnum WHERE a.attnum > 0 AND NOT a.attisdropped) p ON p.nspname=i.table_schema and p.relname=i.table_name and p.attname=i.column_name WHERE i.table_catalog = $1 AND i.table_schema = $2 AND i.table_name = $3", g.database, g.schema, table))
	defer rows.Close()
	for rows.Next() {
		exists = true
		var (
			column      string
			udtName     string
			attndims    int
			description *string
			fieldType   string
		)
		mustNoError(rows.Scan(&column, &udtName, &attndims, &description))
		if description == nil {
			description = gdao.Ptr("")
		}
		if udtName[:1] == "_" {
			udtName = udtName[1:]
		}
		switch udtName {
		case "int8":
			fieldType = "*int64"
		case "int4":
			fieldType = "*int32"
		case "int2":
			fieldType = "*int16"
		case "time", "timetz", "date", "timestamp", "timestamptz":
			fieldType = "*time.Time"
		case "float8", "numeric", "money":
			fieldType = "*float64"
		case "float4":
			fieldType = "*float32"
		case "bool":
			fieldType = "*bool"
		case "bytea":
			fieldType = "[]byte"
		case "bit", "varbit", "box", "bpchar", "varchar", "cidr", "circle", "inet", "interval", "json", "jsonb", "line",
			"lseg", "macaddr", "macaddr8", "path", "pg_lsn", "pg_snapshot", "point", "polygon", "text",
			"tsquery", "tsvector", "txid_snapshot", "uuid", "xml":
			fieldType = "*string"
		}
		for i := 0; i < attndims; i++ {
			if i == 0 {
				fieldType = strings.TrimPrefix(fieldType, "*")
			}
			fieldType = "[]" + fieldType
		}

		f := &field{
			Column:    column,
			FieldName: g.c.FieldNameMapper.Convert(column),
			FieldType: fieldType,
			Comment:   *description,
			Valid:     fieldType != "any",
		}
		fields = append(fields, f)
	}

	rows = mustReturn(g.db.Query("SELECT d.description FROM information_schema.tables i JOIN pg_class c ON c.relname = i.table_name LEFT JOIN pg_description d ON d.objoid = c.oid AND d.objsubid = '0' WHERE i.table_catalog = $1 AND i.table_schema = $2 AND i.table_name = $3", g.database, g.schema, table))
	defer rows.Close()
	if rows.Next() {
		rows.Scan(&tableComment)
	}

	return exists, fields, tableComment
}

func newPostgresGenerator(c Config) postgresGenerator {
	db, err := sql.Open("postgres", c.Dsn)
	if err != nil { // coverage-ignore
		panic(fmt.Sprintf("connect db fail, dsn: %s, error: %v", c.Dsn, err))
	}

	schema := ""
	rows := mustReturn(db.Query("SELECT CURRENT_SCHEMA()"))
	if rows.Next() {
		rows.Scan(&schema)
	}
	rows.Close()

	database := ""
	rows = mustReturn(db.Query("SELECT CURRENT_DATABASE()"))
	if rows.Next() {
		rows.Scan(&database)
	}
	rows.Close()
	return postgresGenerator{c: c, db: db, schema: schema, database: database}
}
