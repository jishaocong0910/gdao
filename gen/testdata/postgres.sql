CREATE TABLE postgres
(
    "bigint"                   BIGINT generated by default as identity,
    "bigserial"                BIGSERIAL,
    "bit"                      BIT(5),
    "bit_arying"               bit varying(255),
    "boolean"                  BOOLEAN[],
    "box"                      box,
    "bytea"                    BYTEA,
    "character"                character(255),
    "character_varying"        character varying(255),
    "cidr"                     cidr,
    "circle"                   circle,
    "date"                     date,
    "double_precision"         double precision,
    "inet"                     inet,
    "integer"                  integer,
    "interval"                 interval,
    "json"                     json,
    "jsonb"                    jsonb,
    "line"                     line,
    "lseg"                     lseg,
    "macaddr"                  macaddr,
    "macaddr8"                 macaddr8,
    "money"                    money,
    "numeric"                  numeric,
    "path"                     path,
    "pg_lsn"                   pg_lsn,
    "pg_snapshot"              pg_snapshot,
    "point"                    point,
    "polygon"                  polygon,
    "real"                     real,
    "smallint"                 smallint,
    "smallserial"              smallserial,
    "serial"                   serial,
    "text"                     text,
    "time"                     time,
    "time_with_time_zone"      time with time zone,
    "timestamp"                timestamp,
    "timestamp_with_time_zone" timestamp with time zone,
    "tsquery"                  tsquery,
    "tsvector"                 tsvector,
    "txid_snapshot"            txid_snapshot,
    "uuid"                     uuid,
    "xml"                      xml
);

COMMENT ON TABLE postgres IS 'postgres';
COMMENT ON COLUMN postgres.bigint IS 'bigint';
COMMENT ON COLUMN postgres.bigserial IS 'bigserial';
COMMENT ON COLUMN postgres.bit IS 'bit';
COMMENT ON COLUMN postgres.bit_arying IS 'bit arying';
COMMENT ON COLUMN postgres.boolean IS 'boolean';
COMMENT ON COLUMN postgres.box IS 'box';
COMMENT ON COLUMN postgres.bytea IS 'bytea';
COMMENT ON COLUMN postgres.character IS 'character';
COMMENT ON COLUMN postgres.character_varying IS 'character varying';
COMMENT ON COLUMN postgres.cidr IS 'cidr';
COMMENT ON COLUMN postgres.circle IS 'circle';
COMMENT ON COLUMN postgres.date IS 'date';
COMMENT ON COLUMN postgres.double_precision IS 'double precision';
COMMENT ON COLUMN postgres.inet IS 'inet';
COMMENT ON COLUMN postgres.integer IS 'integer';
COMMENT ON COLUMN postgres.interval IS 'interval';
COMMENT ON COLUMN postgres.json IS 'json';
COMMENT ON COLUMN postgres.jsonb IS 'jsonb';
COMMENT ON COLUMN postgres.line IS 'line';
COMMENT ON COLUMN postgres.lseg IS 'lseg';
COMMENT ON COLUMN postgres.macaddr IS 'macaddr';
COMMENT ON COLUMN postgres.macaddr8 IS 'macaddr8';
COMMENT ON COLUMN postgres.money IS 'money';
COMMENT ON COLUMN postgres.numeric IS 'numeric';
COMMENT ON COLUMN postgres.path IS 'path';
COMMENT ON COLUMN postgres.pg_lsn IS 'pg_lsn';
COMMENT ON COLUMN postgres.pg_snapshot IS 'pg_snapshot';
COMMENT ON COLUMN postgres.point IS 'point';
COMMENT ON COLUMN postgres.polygon IS 'polygon';
COMMENT ON COLUMN postgres.real IS 'real';
COMMENT ON COLUMN postgres.smallint IS 'smallint';
COMMENT ON COLUMN postgres.smallserial IS 'smallserial';
COMMENT ON COLUMN postgres.serial IS 'serial';
COMMENT ON COLUMN postgres.text IS 'text';
COMMENT ON COLUMN postgres.time IS 'time';
COMMENT ON COLUMN postgres.time_with_time_zone IS 'time with time zone';
COMMENT ON COLUMN postgres.timestamp IS 'timestamp';
COMMENT ON COLUMN postgres.timestamp_with_time_zone IS 'timestamp with time zone';
COMMENT ON COLUMN postgres.tsquery IS 'tsquery';
COMMENT ON COLUMN postgres.tsvector IS 'tsvector';
COMMENT ON COLUMN postgres.txid_snapshot IS 'txid_snapshot';
COMMENT ON COLUMN postgres.uuid IS 'uuid';