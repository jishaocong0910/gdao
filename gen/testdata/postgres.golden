// Code generated by https://github.com/alertego/gdao/gen. YOU CAN EDIT FOR MORE.

package testdata

import "time"

// Postgres postgres
type Postgres struct {
	// Bigint
	Bigint *int64 `gdao:"column=bigint"`
	// Bigserial bigserial
	Bigserial *int64 `gdao:"column=bigserial"`
	// Bit bit
	Bit *string `gdao:"column=bit"`
	// BitArying bit arying
	BitArying *string `gdao:"column=bit_arying"`
	// Boolean boolean
	Boolean []bool `gdao:"column=boolean"`
	// Box box
	Box *string `gdao:"column=box"`
	// Bytea bytea
	Bytea []byte `gdao:"column=bytea"`
	// Character character
	Character *string `gdao:"column=character"`
	// CharacterVarying character varying
	CharacterVarying *string `gdao:"column=character_varying"`
	// Cidr cidr
	Cidr *string `gdao:"column=cidr"`
	// Circle circle
	Circle *string `gdao:"column=circle"`
	// Date date
	Date *time.Time `gdao:"column=date"`
	// DoublePrecision double precision
	DoublePrecision *float64 `gdao:"column=double_precision"`
	// Inet inet
	Inet *string `gdao:"column=inet"`
	// Integer integer
	Integer *int32 `gdao:"column=integer"`
	// Interval interval
	Interval *string `gdao:"column=interval"`
	// Json json
	Json *string `gdao:"column=json"`
	// Jsonb jsonb
	Jsonb *string `gdao:"column=jsonb"`
	// Line line
	Line *string `gdao:"column=line"`
	// Lseg lseg
	Lseg *string `gdao:"column=lseg"`
	// Macaddr macaddr
	Macaddr *string `gdao:"column=macaddr"`
	// Macaddr8 macaddr8
	Macaddr8 *string `gdao:"column=macaddr8"`
	// Money money
	Money *float64 `gdao:"column=money"`
	// Numeric numeric
	Numeric *float64 `gdao:"column=numeric"`
	// Path path
	Path *string `gdao:"column=path"`
	// PgLsn pg_lsn
	PgLsn *string `gdao:"column=pg_lsn"`
	// PgSnapshot pg_snapshot
	PgSnapshot *string `gdao:"column=pg_snapshot"`
	// Point point
	Point *string `gdao:"column=point"`
	// Polygon polygon
	Polygon *string `gdao:"column=polygon"`
	// Real real
	Real *float32 `gdao:"column=real"`
	// Smallint smallint
	Smallint *int16 `gdao:"column=smallint"`
	// Smallserial smallserial
	Smallserial *int16 `gdao:"column=smallserial"`
	// Serial serial
	Serial *int32 `gdao:"column=serial"`
	// Text text
	Text *string `gdao:"column=text"`
	// Time time
	Time *time.Time `gdao:"column=time"`
	// TimeWithTimeZone time with time zone
	TimeWithTimeZone *time.Time `gdao:"column=time_with_time_zone"`
	// Timestamp timestamp
	Timestamp *time.Time `gdao:"column=timestamp"`
	// TimestampWithTimeZone timestamp with time zone
	TimestampWithTimeZone *time.Time `gdao:"column=timestamp_with_time_zone"`
	// Tsquery tsquery
	Tsquery *string `gdao:"column=tsquery"`
	// Tsvector tsvector
	Tsvector *string `gdao:"column=tsvector"`
	// TxidSnapshot txid_snapshot
	TxidSnapshot *string `gdao:"column=txid_snapshot"`
	// Uuid uuid
	Uuid *string `gdao:"column=uuid"`
	// Xml xml
	Xml *string `gdao:"column=xml"`
}
